lib/compat/wordpress-5.9/block-template-utils.php

```diff
@@ -755,33 +755,20 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
 	if ( count( $parts ) < 2 ) {
 		return null;
 	}
-	list( $theme, $slug ) = $parts;
-	$wp_query_args        = array(
-		'post_name__in'  => array( $slug ),
-		'post_type'      => $template_type,
-		'post_status'    => array( 'auto-draft', 'draft', 'publish', 'trash' ),
-		'posts_per_page' => 1,
-		'no_found_rows'  => true,
-		'tax_query'      => array(
-			array(
-				'taxonomy' => 'wp_theme',
-				'field'    => 'name',
-				'terms'    => $theme,
-			),
-		),
-	);
-	$template_query       = new WP_Query( $wp_query_args );
-	$posts                = $template_query->posts;
-
-	if ( count( $posts ) > 0 ) {
-		$template = _build_block_template_result_from_post( $posts[0] );
-
-		if ( ! is_wp_error( $template ) ) {
-			return $template;
-		}
+	list( , $slug ) = $parts;
+	$block_template = resolve_block_template( $slug, array(), '' );
+	if ( ! $block_template ) {
+		$block_template = resolve_block_template( 'index', array(), '' );
+	}
+	// This might give us a fallback template with a different ID,
+	// so we have to override it to make sure it's correct.
+	$block_template->id     = $id;
+	$block_template->slug   = $slug;
+	$default_template_types = get_default_block_template_types();
+	if ( array_key_exists( $slug, $default_template_types ) ) {
+		$block_template->title       = $default_template_types[ $slug ]['title'];
+		$block_template->description = $default_template_types[ $slug ]['description'];
 	}
-
-	$block_template = get_block_file_template( $id, $template_type );
```

Comment:
Personally I feel the logic change here doesn't make sense, at least not from the `gutenberg_get_block_template` perspective, this is a low level API to retrieve a template given its ID. Potentially it can be moved to the endpoint to create a template but not while quering. WDYT

Reply:
Yeah, that's the big question here. I think that there's good arguments for each side: 

- It _is_ indeed a fairly low-level API, and it seems like when given a template ID, it should just return that template (or nothing, if it can't find it).
- OTOH, I think a point can be made that we've now identified one significant use case where that doesn't cut it, and we need the fallback. My hypothesis was that this could be the case pretty much anywhere where we'd call `gutenberg_get_block_template` (or "require a template given its ID") -- not just for the endpoint -- since it replicates what happens on the frontend, and it's likely that we _always_ want to replicate that.

I'll do a quick audit of all the other occurrences of `gutenberg_get_block_template`. If indeed I find that the fallback makes sense in all current cases, I'll probably opt to change it to that effect. (The argument then being YAGNI: If we currently only need it _with_ the fallback, we should only provide it that way.)

Another option would be to add a bool arg to the function signature to indicate whether or not to look for fallback templates or not, but I'd only do that if we find that we really need both versions of the function already.

Reply:
> I'll do a quick audit of all the other occurrences of `gutenberg_get_block_template`.

https://github.com/WordPress/gutenberg/pull/37258#issuecomment-990895049